src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,405 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief A reverse-mode automatic differentiation.
+ *
+ * Now only supports differentiating a function in the IRModule with one dataflow block
+ * with respect to the only return value of the function. It needs to be scalar.
+ *
+ * For examples, see the MLP examples in tests/python/relax/test_transform_gradient.py and
+ * tests/python/relax/test_transform_gradient_numeric.py.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+namespace tvm {
+namespace relax {
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(IRModule mod, const Array<Var>& require_grads)
+      : ExprMutator(mod), require_grads_(require_grads) {}
+
+  Function FuncTransform(const FunctionNode* node) {
```

Comment:
Usually we pass references instead of pointers in APIs. And better naming as well.
```suggestion
  Function FuncTransform(Function func) {
```